shell: Add defaultDexConnector option to automatically trigger login on a predefined Dex connector

JBWatenbergScality opened the pull request:
**Component**:

shell

**Context**: 

Dex is providing the possibility to use several connectors to federate OIDC providers. However under some conditions we may need to force the usage of a specific connector.

**Summary**:

This PR add a `defaultDexConnector` option to enable this.

This option can be added to the auth section of micro application `runtime-app-configuration`.
It intercept the `authorization_endpoint` of the oidc web finger and append it the `connector_id` query parameter.

The connector choice can be forced by adding the `displayLoginChoice` query parameter to the current URL.
This enable to maintain UI login possible when the default connector is unavailable.

Additionaly as Dex is not supporting `end_session_endpoint` the logout is reviewed to clean the cookies of the
domain and clearing the localStorage entries of oidc client when `providerLogout` is set to true.

**Acceptance criteria**: 

We shall be able to login directly using the default dex connector when the property is defined

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-dex-default-connector-option` into
`development/2.10`
* [w/2.11/improvement/add-dex-default-connector-option](https://github.com/scality/metalk8s/compare/development/2.11...w/2.11/improvement/add-dex-default-connector-option) will be merged into `development/2.11` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan

ChengYanJin commented on shell-ui/src/auth/AuthProvider.js:
```diff
@@ -39,6 +57,9 @@ function OAuth2AuthProvider({ children }: { children: Node }) {
     loadUserInfo: true,
     automaticSilentRenew: true,
     monitorSession: false,
+    MetadataServiceCtor: authConfig.defaultDexConnector
```
I am wondering should we add this `defaultDexConnector` prop to type `OIDCConfig`?

JBWatenbergScality replied:
Done in db1812c1d0d9a5d1e21a20caefd003b31b9a1a75

ChengYanJin commented on shell-ui/src/auth/AuthProvider.js:
```diff
@@ -39,6 +57,9 @@ function OAuth2AuthProvider({ children }: { children: Node }) {
     automaticSilentRenew: true,
     monitorSession: false,
+    MetadataServiceCtor: authConfig.defaultDexConnector
+      ? defaultDexConnectorMetadataService(authConfig.defaultDexConnector)
```
It's more a question: 
If we want to use Dex to authenticate the users against a connector-based interface, how should we set the `defaultDexConnector` in `runtime-app-configuration`?

gdemonet replied:
Didn't we start documenting all this rework of shell-ui? Would be useful IMO, both now for reviewers catching up with the progress, and in the future for maintenance / onboarding.

JBWatenbergScality replied:
`defaultDexConnector` should take the connector id as defined in dex configuration.
For example given the following Dex connectors configuration :
```
       connectors:
            - type: oidc
              id: keycloak
              name: Keycloak
              config:
                   //....
```

Providing `keycloak` as `defaultDexConnector` in `runtime-app-configuration` will automatically redirect the user to this connector.

JBWatenbergScality replied:
I've opened #3476 regarding the documentation and will address it as soon as possible

gdemonet commented on shell-ui/src/auth/AuthProvider.js:
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
+              console.log(
+                "Oidc provider doesn't support end session endpoint, fallback to clearing document cookies",
+              );
+              document.cookie.split(';').forEach(function (c) {
+                document.cookie =
+                  c.trim().split('=')[0] +
+                  '=;' +
+                  'expires=Thu, 01 Jan 1970 00:00:00 UTC;';
+              });
```
What's that? We split the cookie and assign a new value for each? How does it work? Is it that assigning a string to `document.cookie` is actually interpreted as an instruction, rather than describing the whole "cookies state"?

JBWatenbergScality replied:
Yes using `document.cookie` we can only set or update a single cookie at a time (https://developer.mozilla.org/en-US/docs/web/api/document/cookie#write_a_new_cookie). What I'm trying to achieve here is a hack to logout the users from the federated OIDC Provider behind Dex because Dex is not supporting the end session endpoint and then the session on the original OIDC provider can't be closed in any other way 😕 .

gdemonet commented on shell-ui/src/auth/AuthProvider.js:
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
+          auth.userManager.signoutRedirect().catch((e) => {
+            if (e.message === 'no end session endpoint') {
+              console.log(
+                "Oidc provider doesn't support end session endpoint, fallback to clearing document cookies",
```
```suggestion
                "OIDC provider doesn't support end session endpoint, fallback to clearing document cookies",
```

JBWatenbergScality replied:
Done in e6566809c79b0f9ea3cb11fa81d856096bb04d06

gdemonet commented on shell-ui/src/auth/AuthProvider.js:
```diff
@@ -95,28 +116,45 @@ export function useLogOut() {
+            }
+            auth.userManager.removeUser();
+            location.reload();
+          });
```
This whole `.catch` block doesn't seem related to the change in this PR (which is about a pre-defined Dex connector). Would be nice to at least use a different commit :wink:

JBWatenbergScality replied:
Done in e6566809c79b0f9ea3cb11fa81d856096bb04d06

gdemonet reviewed (changes requested):
No need to tackle documentation as part of this PR :wink:

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

JBWatenbergScality committed:
shell: Add defaultDexConnector option to automatically trigger login on a predefined Dex connector.

This option can be added to the auth section of micro application `runtime-app-configuration`.
It intercept the authorization_endpoint of the oidc web finger and append it the `connector_id` query parameter.

The connector choice can be forced by adding the `displayLoginChoice` query parameter to the current URL.
This enable to maintain UI login possible when the default connector is unavailable.

JBWatenbergScality committed:
shell: Review logout to clean the cookies of the domain when the end_session_endpoint is not supported

Dex is not supporting `end_session_endpoint` the logout is reviewed to clean the cookies of the
domain and clearing the localStorage entries of oidc client when `providerLogout` is set to true.

Ref: #2831

bert-e commented:
# History mismatch



Merge commit #72d160dc1c6470be75c092cfd23c7ed969af5b1f on the integration branch
`w/2.11/improvement/add-dex-default-connector-option` is merging a branch which is neither the current
branch `improvement/add-dex-default-connector-option` nor the development branch
`development/2.11`.

It is likely due to a rebase of the branch `improvement/add-dex-default-connector-option` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-dex-default-connector-option` into
`development/2.10`
* [w/2.11/improvement/add-dex-default-connector-option](https://github.com/scality/metalk8s/compare/development/2.11...w/2.11/improvement/add-dex-default-connector-option) will be merged into `development/2.11` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/add-dex-default-connector-option` has diverged from
`development/2.11` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.11` into `improvement/add-dex-default-connector-option`
* Rebase `improvement/add-dex-default-connector-option` onto `origin/development/2.11`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/add-dex-default-connector-option` has diverged from
`development/2.11` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.11` into `improvement/add-dex-default-connector-option`
* Rebase `improvement/add-dex-default-connector-option` onto `origin/development/2.11`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @teddyandrieux

* @thomasdanan

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @gdemonet

* @alexandre-allard-scality

JBWatenbergScality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye jbwatenbergscality.
